libcudacxx/include/cuda/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,431 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA__MEMORY_RESOURCE_CUDA_MEMORY_POOL_H
+#define _CUDA__MEMORY_RESOURCE_CUDA_MEMORY_POOL_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// cudaMallocAsync was introduced in CTK 11.2
+#if !defined(_CCCL_COMPILER_MSVC_2017) && !defined(_CCCL_CUDACC_BELOW_11_2)
+
+#  if !defined(_CCCL_CUDA_COMPILER_NVCC) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
+#    include <cuda_runtime.h>
+#    include <cuda_runtime_api.h>
+#  endif // !_CCCL_CUDA_COMPILER_NVCC && !_CCCL_CUDA_COMPILER_NVHPC
+
+#  include <cuda/__memory_resource/get_property.h>
+#  include <cuda/__memory_resource/properties.h>
+#  include <cuda/__memory_resource/resource.h>
+#  include <cuda/__memory_resource/resource_ref.h>
+#  include <cuda/std/__cuda/api_wrapper.h>
+#  include <cuda/std/__new/bad_alloc.h>
+#  include <cuda/stream_ref>
+
+#  if _CCCL_STD_VER >= 2014
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_MR
+
+/**
+ * @brief Internal redefinition of cudaMemAllocationHandleType
+ *
+ * @note We need to define our own enum here because the earliest CUDA runtime version that supports asynchronous
+ * memory pools (CUDA 11.2) did not support these flags. See the `cudaMemAllocationHandleType` docs at
+ * https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html
+ */
+enum class cudaMemAllocationHandleType
+{
+  cudaMemHandleTypeNone                = 0 << 0, ///< Does not allow any export mechanism.
+  cudaMemHandleTypePosixFileDescriptor = 0 << 1, ///< Allows a file descriptor to be used for exporting
+  cudaMemHandleTypeWin32               = 0 << 2, ///< Allows a Win32 NT handle to be used for exporting. (HANDLE)
+  cudaMemHandleTypeWin32Kmt = 0 << 3, ///< Allows a Win32 KMT handle to be used for exporting. (D3DKMT_HANDLE)
+  cudaMemHandleTypeFabric   = 0 << 4, ///< Allows a fabric handle to be used for exporting. (cudaMemFabricHandle_t)
+};
+
+/**
+ * @brief `cuda_memory_pool` uses cudaMallocAsync / cudaFreeAsync for allocation/deallocation.
+ *
+ * See https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html
+ */
+class cuda_memory_pool
```

Comment:
Technically, a `cuda_memory_pool` satisfies all the `cuda::mr::memory_resource*` concepts, and implements a `get_property()`. So it is a memory resource. Based on that, I slightly lean toward naming it as such.

Also the name is ambiguous -- anyone's device memory pool resource could be called `cuda_memory_pool_resource`.

So I guess what I'm saying it... more bikeshedding is warranted. :)

Reply:
How about `cuda_owning_asnyc_memory_resource`

Reply:
See later comment. https://github.com/NVIDIA/cccl/pull/1637#issuecomment-2073641128

Reply:
We decided to turn `cuda_memory_pool` into a thin wrapper around `cudaMemPool_t`